src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -267,23 +275,142 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudInstanceId = result.getCloudInstanceId();
+        if (cloudInstanceId == null || cloudInstanceId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud instance ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isBlank() && result.isEntitled()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
+            }
+        }
+
+        String anonymousConsumerUuid = null;
+        if (ownerKey == null || ownerKey.isBlank() || !result.isEntitled() || !ownerAndPoolExists) {
+
+            // TODO: Schedule anonymous org creation job
+
+            AnonymousCloudConsumer anonConsumer = anonymousCloudConsumerCurator
+                .getByCloudInstanceId(cloudInstanceId);
+
+            if (anonConsumer == null) {
+                anonConsumer = new AnonymousCloudConsumer()
+                    .setCloudAccountId(cloudAccountId)
+                    .setCloudInstanceId(cloudInstanceId)
+                    .setProductId(productId)
+                    .setCloudProviderShortName(cloudProviderShortName);
+
+                anonConsumer = anonymousCloudConsumerCurator.create(anonConsumer);
+            }
+
+            anonymousConsumerUuid = anonConsumer.getUuid();
+        }
+
+        String tokenType = anonymousConsumerUuid == null ? TOKEN_TYPE : ANON_TOKEN_TYPE;
+        String token = anonymousConsumerUuid == null ?
+            this.buildRegistrationToken(principal, ownerKey, tokenType) :
+            this.buildRegistrationToken(principal, "", tokenType);
```

Comment:
One thing that's missing in the anonymous token use case: the expiration of the anonymous token should be set at 2 days; not at whatever the current token is using. The current token has a config property (ConfigProperties.JWT_TOKEN_TTL) that gets fed into this class. Perhaps we need a separate property called something like ANON_JWT_TOKEN_TTL, set to a default of 172800 seconds (2 days).

Reply:
Added